### PR TITLE
Polish variant-aware dependency substitution

### PR DESCRIPTION
**Pull request by melix**
This PR builds on top of #13299  and adds documentation and missing APIs.

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test RFM

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyState.java [hunk 1]
```diff
@@ -33,10 +33,7 @@
 import java.util.List;
 import java.util.stream.Collectors;
 
-import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ComponentSelectionReasons.BY_ANCESTOR;
-import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ComponentSelectionReasons.CONSTRAINT;
-import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ComponentSelectionReasons.FORCED;
-import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ComponentSelectionReasons.REQUESTED;
+import static org.gradle.api.internal.artifacts.ivyservice.resolveengine.result.ComponentSelectionReasons.*;
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyState.java:36]: This looks like an unintended change, and not aligned with usual code style

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -74,6 +74,16 @@ main(MavenPublication) {
 
 See the documentation for more information on [Gradle Module Metadata generation](userguide/publishing_gradle_module_metadata.html#sub:gmm-reproducible).
 
+## Variant-aware dependency substitution rules
+
+Gradle now supports substitution of dependencies with classifiers, attributes or capabilities.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/release/notes.md:79]: I would move that sentence below the part that explains the previous limitations.

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -74,6 +74,16 @@ main(MavenPublication) {
...
+Previously, it wasn't possible for Gradle to substitute a dependency which uses a classifier with a dependency that didn't, nor was it possible to substitute a dependency _without_ classifier with a classified dependency.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/release/notes.md:80]: I find the following part hard to read. I would be fully explicit instead.

> with a dependency that didn't.

File: subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/resolution_rules.adoc [hunk 4]
```diff
@@ -183,6 +183,81 @@ include::sample[dir="snippets/dependencyManagement/customizingResolution-conditi
...
 Dependency substitution rules take care of replacing the module dependency with the project dependency, but do not implicitly include the project in the build.
 
+[[sec:variant_aware_substitutions]]
+=== Substituting a dependency with another variant
+
+Gradle's dependency management engine is <<variant_model.adoc#understanding-variant-selection,variant-aware>> meaning that for a single component, the engine may select different artifacts and transitive dependencies.
+
+What to select is determined by the attributes of the consumer configuration and the attributes of the variants found on the producer side.
+It is, however, possible that some specific dependencies override attributes from the configuration itself.
+This is typically the case when using the <<java_platform_plugin.adoc#java_platform_plugin,Java Platform plugin>>: this plugin builds a special kind of component which is called a "platform" and can be addressed by setting the component category attribute to "platform", in opposition to typical dependencies which are targetting libraries.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/resolution_rules.adoc:193]: I would use code font (with backticks) for the values of the category attribute.

**Pull request merged**